### PR TITLE
Get events unit tests

### PR DESCRIPTION
**Pull request by apoorvsadana**
# Pull Request type

Please add the labels corresponding to the type of changes your PR introduces:

- Testing

## What is the current behavior?

We only have integration tests for get_events. So we need to run the client, add blocks, call get_events and check if it's working as expected.

Resolves: #517 

## What is the new behavior?

Unit tests have been added and the test structure has been setup so that we can now test rpc methods and internal functions without having to run the client.

## Does this introduce a breaking change?

No

### CODE REVIEW

File: crates/client/rpc/src/lib.rs [hunk 1]
```diff
@@ -150,7 +153,85 @@ where
...
+    /// * `address` - Address to use to filter the events
+    /// * `keys` - Keys to use to filter the events. An event is filtered if any key is present
+    /// * `max_results` - Optional, indicated the max events that need to be filtered
+    ///
+    /// # Returns
+    ///
+    /// * `(block_events: Vec<EventWrapper>, continuation_token: usize)` - A tuple of the filtered
+    ///   events and the first index which still hasn't been processed block_id and an instance of
+    ///   Block
+    pub fn filter_events_by_params(
```
Comment (tdelabro) [inline crates/client/rpc/src/lib.rs:202]: This does not need to be part of a `impl` block.
It does not use `self` at all. It can just be a regular function.
This way you can test it without building the `Starknet` type.

And you can test it here, in `client/rpc`, rather than in `node`

File: crates/client/rpc/src/lib.rs [hunk 2]
```diff
@@ -207,36 +264,42 @@ where
                 continue;
             }
 
-            let block_events = block_events.into_iter().skip(continuation_token);
+            let block_events = block_events.into_iter().skip(continuation_token).collect();
```
Comment (tdelabro) [inline crates/client/rpc/src/lib.rs:267]: You are collecting twice, which is a waste we tried to avoid when we originally designed this method

File: crates/node/Cargo.toml [hunk 3]
```diff
@@ -83,6 +84,10 @@ blockifier = { workspace = true }
 hex = { workspace = true }
 starknet-core = { workspace = true }
 
+# Testing
+tokio = "1.28.2"
+serial_test = "2.0.0" # to run rpc tests sequentially, else all tests try to call rpc methods at the same time and we get lock errors
+
```
Comment (tdelabro) [inline crates/node/Cargo.toml:90]: Testing import should go into `[dev-dependencies]`

**Review by tdelabro — CHANGES_REQUESTED** [review]

File: crates/client/rpc/src/lib.rs (hunk 1 above)
Comment (apoorvsadana) [reply crates/client/rpc/src/lib.rs]: moved it outside the impl block and move the test cases within rpc

File: crates/client/rpc/src/lib.rs (hunk 2 above)
Comment (apoorvsadana) [reply crates/client/rpc/src/lib.rs]: removed the collect here and changed the argument type of the filter function

File: crates/client/rpc/src/lib.rs [hunk 4]
```diff
@@ -937,3 +962,47 @@ where
...
+/// * `events` - A vector of all events
+/// * `address` - Address to use to filter the events
+/// * `keys` - Keys to use to filter the events. An event is filtered if any key is present
+/// * `max_results` - Optional, indicated the max events that need to be filtered
+///
+/// # Returns
+///
+/// * `(block_events: Vec<EventWrapper>, continuation_token: usize)` - A tuple of the filtered
+///   events and the first index which still hasn't been processed block_id and an instance of Block
+pub fn filter_events_by_params(
```
Comment (tdelabro) [inline crates/client/rpc/src/lib.rs:979]: move it in a different files `events.rs`
Along with the `get_block_events` method and basically anything event-related you can find in the file.

It will help readablility and code maintaining

File: crates/client/rpc/src/lib.rs [hunk 5]
```diff
@@ -937,3 +962,47 @@ where
...
+    keys: Vec<Vec<FieldElement>>,
+    max_results: Option<usize>,
+) -> (Vec<EventWrapper>, usize) {
+    let mut filtered_events = vec![];
+    let mut index = 0;
+
+    // Iterate on block events.
+    for event in events {
+        index += 1;
+        let match_from_address = address.map_or(true, |add| add == event.from_address);
```
Comment (tdelabro) [inline crates/client/rpc/src/lib.rs:991]: `add` is a bad abreviation for address. It sound like addition
addr is better

File: crates/client/rpc/src/lib.rs [hunk 6]
```diff
@@ -150,7 +158,39 @@ where
...
+        let block = get_block_by_block_hash(self.client.as_ref(), substrate_block_hash).ok_or_else(|| {
+            error!("Failed to retrieve block");
+            StarknetRpcApiError::BlockNotFound
+        })?;
+        let block_events = self
+            .overrides
+            .for_block_hash(self.client.as_ref(), substrate_block_hash)
+            .events(substrate_block_hash)
+            .unwrap_or_else(|| {
+                info!("No events found in block {}", block_id);
```
Comment (tdelabro) [inline crates/client/rpc/src/lib.rs:187]: info! is probably too strong.
dbg! should be enough

**Review by tdelabro — CHANGES_REQUESTED** [review]

File: crates/client/rpc/Cargo.toml [hunk 7]
```diff
@@ -52,3 +52,7 @@ jsonrpsee = { workspace = true, default-features = true, features = [
 log = { workspace = true, default-features = true }
 serde_json = { workspace = true, default-features = true }
 thiserror = { workspace = true }
+
+[dev-dependencies]
+rstest = { workspace = true, default-features = true }
```
[the remark this answers is not in the document]
Comment (apoorvsadana) [reply crates/client/rpc/Cargo.toml]: removed default-features from here

File: crates/client/rpc/src/lib.rs (hunk 4 above)
Comment (apoorvsadana) [reply crates/client/rpc/src/lib.rs]: moved `filter_events_by_params` to a new file, however `get_block_events` depends on `self`, so should that stay inside the implementation

File: crates/client/rpc/src/lib.rs (hunk 5 above)
Comment (apoorvsadana) [reply crates/client/rpc/src/lib.rs]: fixed

File: crates/client/rpc/src/tests.rs [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+	_name: &'a str,
+	events:Vec<EventWrapper>,
+	filter_keys: Vec<Vec<FieldElement>>,
+	filter_address: Option<Felt252Wrapper>,
+	max_results: Option<usize>,
+	expected_events: Vec<EventWrapper>,
+	expected_continuation_token: usize,
+}
+
+#[fixture]
```
[the remark this answers is not in the document]
Comment (apoorvsadana) [reply crates/client/rpc/src/tests.rs:21]: added `#[once]`

File: crates/client/rpc/src/lib.rs (hunk 6 above)
Comment (apoorvsadana) [reply crates/client/rpc/src/lib.rs]: fixed

File: crates/client/rpc/src/lib.rs (hunk 4 above)
Comment (tdelabro) [reply crates/client/rpc/src/lib.rs]: you can have multiple `impl MyType {}` blocks

Comment (apoorvsadana) [reply crates/client/rpc/src/lib.rs]: moved to `events.rs`

**Comment by abdelhamidbakhta** [comment]
@tdelabro can you do a final review please ? @apoorvsadana applied your suggestions

**Comment by apoorvsadana** [comment]
there was a conflict with `.gitignore` after the recent addition of `.vscode`, just resolved that

File: crates/client/rpc/src/events.rs [hunk 9]
```diff
@@ -1,10 +1,66 @@
...
+use crate::Starknet;
+
+impl<B, BE, C, P, H> Starknet<B, BE, C, P, H>
+where
+    B: BlockT,
+    C: HeaderBackend<B> + StorageProvider<B, BE> + 'static,
+    C: ProvideRuntimeApi<B>,
+    C::Api: StarknetRuntimeApi<B> + ConvertTransactionRuntimeApi<B>,
+    BE: Backend<B>,
+    H: HasherT + ThreadSafeCopy,
```
Comment (tdelabro) [inline crates/client/rpc/src/events.rs:29]: Are you sure that everything here is absolutely necessary?
Did you try removing some of those

File: crates/client/rpc/src/lib.rs [hunk 10]
```diff
@@ -157,38 +154,6 @@ where
...
-            .overrides
-            .for_block_hash(self.client.as_ref(), substrate_block_hash)
-            .events(substrate_block_hash)
-            .unwrap_or_else(|| {
-                dbg!("No events found in block {}", block_id);
-                Vec::new()
-            });
-        Ok((block_events, block))
-    }
-
```
Comment (tdelabro) [inline crates/client/rpc/src/lib.rs]: `filter_events` can also be moved to the events.rs file

File: crates/client/rpc/src/tests.rs [hunk 11]
```diff
@@ -0,0 +1,125 @@
+use mp_starknet::execution::types::{ContractAddressWrapper, Felt252Wrapper};
```
Comment (tdelabro) [inline crates/client/rpc/src/tests.rs:1]: the file hierary could be:
src/
   events/
     mod.rs
     tests.rs
     
     
Because all the tests in this file are event related

File: crates/client/rpc/src/tests.rs [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+}
+
+#[rstest]
+#[case::filter_keys(build_test_case()[0].clone())]
+#[case::filter_address(build_test_case()[1].clone())]
+#[case::filters_keys_and_address(build_test_case()[2].clone())]
+#[case::filter_max_results_met(build_test_case()[3].clone())]
+#[case::filter_max_results_not_met(build_test_case()[4].clone())]
+#[case::filter_keys_less_than_actual(build_test_case()[5].clone())]
+#[case::filter_keys_more_than_actual(build_test_case()[6].clone())]
```
Comment (tdelabro) [inline crates/client/rpc/src/tests.rs:105]: There is no other edge case you could think of?
Better too much that not enough

**Review by tdelabro — CHANGES_REQUESTED** [review]

File: crates/client/rpc/src/events.rs (hunk 9 above)
Comment (apoorvsadana) [reply crates/client/rpc/src/events.rs:29]: Yes, I just checked again as well, everything is necessary

File: crates/client/rpc/src/tests.rs (hunk 11 above)
Comment (apoorvsadana) [reply crates/client/rpc/src/tests.rs]: makes sense! fixed.

File: crates/client/rpc/src/lib.rs (hunk 10 above)
Comment (apoorvsadana) [reply crates/client/rpc/src/lib.rs]: moved

File: crates/client/rpc/src/tests.rs (hunk 12 above)
Comment (apoorvsadana) [reply crates/client/rpc/src/tests.rs]: I thought a little more, I have added two new test cases

- Filter events when there are no filters (return all events)
- Filter events when there are no events (return empty vector)

Apart from these, there are few test cases which I added before, like

- filter_address_chunk_size_met
- filter_address_chunk_size_not_met
- filter_keys_chuck_size_met
- filter_keys_chunk_size_not_met

These felt redundant and I thought it made it challenging to filter the useful cases, so I removed them. However, if you feel we should keep these to protect code changes later, I can add them back.
